### PR TITLE
[Merged by Bors] - feat(Algebra/Order): rewrite positive cones to use substructure pattern

### PR DESCRIPTION
**Pull request by artie2000**
* Rewrite (positive) cones in groups and rings to extend `Submonoid` and `Subsemiring`, respectively. This makes them easier to work with in applications.
* Add multiplicative versions of cones in groups.
* Add `nonneg` constructor for cones from ordered groups/rings.
* Weaken definition of cones in rings slightly (previous definition was nonstandard due to nonstandard definition of linearly ordered ring) -- to compensate, a new fact (ordered domain is strictly ordered domain) has been added.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:97]: This should be `PositiveCone.nonneg`

Comment (FM22) [reply Mathlib/Algebra/Order/Group/Cone.lean:97]: Sorry, I don't understand -- do you mean that I should rename the function?

**Comment by FM22** [comment]
This PR slightly weakens the mathematical definition of a positive cone in a ring, to be in line with the standard definition. The new definition is equivalent to `OrderedRing` rather than `StrictOrderedRing`. To compensate, I have added a new constructor for `StrictOrderedRing` when the underlying ring is a domain.

**Comment by FM22** [comment]
Not adding any more content to this PR -- should all work now.

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:25]:
```suggestion
/-- `PositiveConeClass S G` says that `S` is a type of positive cones in `G`. -/
```
to avoid confusing the mathlib type with the mathematical notion

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:28]:
```suggestion
  eq_zero_of_mem_of_neg_mem {C : S} {a : G} : a ∈ C → -a ∈ C → a = 0
```
if you want

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:34]:
```suggestion

This is equivalent to being the set of non-negative elements of some order on the group making it
into an  `OrderedAddCommGroup`. -/
```

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:48]: Same here

File: Mathlib/Algebra/Order/Ring/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Ring/Cone.lean:42]:
```suggestion
theorem PositiveConeClass.neg_one_not_mem [Nontrivial R] [PositiveConeClass S R] : -1 ∉ C :=
```

**Review by YaelDillies — COMMENTED** [review]
I am not a huge fan of the `AddCommGroup.PositiveCone`/`Ring.PositiveCone` naming convention. Do you think you could rename them to `PositiveGroupCone`/`PositiveRingCone`?

**Comment by FM22** [comment]
> I am not a huge fan of the `AddCommGroup.PositiveCone`/`Ring.PositiveCone` naming convention. Do you think you could rename them to `PositiveGroupCone`/`PositiveRingCone`?

These names are getting absurdly long. Maybe GroupCone and RingCone? With Positive removed everywhere?

Also should I add a multiplicative version for groups? Is there a way to automatically go between them?

**Comment by YaelDillies** [comment]
> Maybe GroupCone and RingCone? With Positive removed everywhere?

Yeah actually that sounds fine. As discussed in private, multiplicative would be nice to have but not necessary

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:48]: Same here

**Review by YaelDillies — APPROVED** [review]
Thanks! Please remember to update the PR description to mention that you are defining `RingConeWithSquares`

maintainer merge

**Comment by FM22** [comment]
> Thanks! Please remember to update the PR description to mention that you are defining `RingConeWithSquares`
> 
> maintainer merge

Where is my PR description? Is it just the very top comment?

File: Mathlib/Algebra/Order/Ring/Cone.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Ring/Cone.lean:29]: Can you extend `AddGroupCone` here too?

Comment (FM22) [reply Mathlib/Algebra/Order/Ring/Cone.lean]: Done

**Comment by FM22** [comment]
I don't understand the error.

**Comment by FM22** [comment]
@YaelDillies could you help if you get a chance? I don't know what the missing docstring is.

**Comment by FM22** [comment]
Putting this on hold because there are a few definitions of cones kicking around and I want to get it right for applications the first time.

**Comment by YaelDillies** [comment]
What about you make this PR do what it says on the tin? namely to rewrite the existing cone structures to use the subobject class pattern, and move the new content to a new PR?

**Comment by FM22** [comment]
Good idea! I will move the squares stuff and the new lemmas to a new PR.

**Comment by FM22** [comment]
I would particularly appreciate feedback on the new `IsMaxCone` design.

File: Mathlib/Algebra/Order/Group/Cone.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Cone.lean:12]:
```suggestion
In this file we provide the structure `GroupCone` and the predicate `IsMaxCone`
```

Comment (FM22) [reply Mathlib/Algebra/Order/Group/Cone.lean]: fixed

**Comment by YaelDillies** [comment]
This looks like a much cleaner diff now.

maintainer merge

File: Mathlib/Algebra/Order/Ring/Basic.lean
Comment (urkud) [inline Mathlib/Algebra/Order/Ring/Basic.lean:126]: You only need `NoZeroDivisors α` here. Also, should it be a low priority instance?

File: Mathlib/Algebra/Order/Ring/Cone.lean
Comment (urkud) [inline Mathlib/Algebra/Order/Ring/Cone.lean:76]: Please add a warning saying that using this as a constructor in an instance can lead to diamonds due to non-customizable `lt`, `decidableLT`, `decidableEq`, and `compare` fields.

**Comment by urkud** [comment]
I see that you add lots of transitive imports to reuse `Submonoid.nonneg`. Should we move these defs to new files?

**Comment by YaelDillies** [comment]
I've opened #16381 to remove some imports to `Algebra.Order.Monoid.Submonoid`

File: Mathlib/Algebra/Order/Ring/Cone.lean
Comment (FM22) [reply Mathlib/Algebra/Order/Ring/Cone.lean:76]: Done -- does this also apply to the previous `def`?

File: Mathlib/Algebra/Order/Ring/Basic.lean
Comment (FM22) [reply Mathlib/Algebra/Order/Ring/Basic.lean:127]: Done the first part. @YaelDillies recommended to me to not make this an instance, due to potential for looping (if I understand right).

**Comment by YaelDillies** [comment]
> I see that you add lots of transitive imports to reuse `Submonoid.nonneg`. Should we move these defs to new files?

The remainder of the imports don't really seem worth eliminating: They are very basic algebra or very basic order theory which simply haven't been needed for cones yet because we don't have much theory.

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**
